### ISSUE TITLE
Panic at rend3::util::registry::archetypical::ArchetypicalRegistry<K,V,HandleType>::get_value_mut

### ISSUE THREAD

**Issue by John-Nagle** [BVE-Reborn/rend3#465]
```
thread 'main' panicked at 'no entry found for key', /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/a972b0f/rend3/src/util/registry/archetypical.rs:126:28
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_display
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:72:5
   3: core::panicking::panic_str
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:56:5
   4: core::option::expect_failed
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:1874:5
   5: rend3::util::registry::archetypical::ArchetypicalRegistry<K,V,HandleType>::get_value_mut
   6: rend3::renderer::ready::ready
   7: rend3::renderer::Renderer::ready
   8: <sharpview::AppUi as rend3_framework::App>::handle_event
   9: rend3_framework::async_start::{{closure}}::{{closure}}
  10: winit::platform_impl::platform::x11::EventLoop<T>::run
  11: winit::platform_impl::platform::EventLoop<T>::run
  12: winit::event_loop::EventLoop<T>::run
  13: rend3_framework::winit_run
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  15: rend3_framework::start
  16: sharpview::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
Never saw this before.

My own code is 100% safe Rust. Recent changes made it more concurrent, and it may be hitting Rend3 with different timing.

Rend3 version info:
```
rend3 = { git = 'https://github.com/BVE-Reborn/rend3.git', rev = "a972b0ff0a98bb79b2b1c8812e3a823840a5e02e" }
rend3-routine = { git = 'https://github.com/BVE-Reborn/rend3.git', rev = "a972b0ff0a98bb79b2b1c8812e3a823840a5e02e" }
rend3-gltf = { git = 'https://github.com/BVE-Reborn/rend3.git', rev = "a972b0ff0a98bb79b2b1c8812e3a823840a5e02e" }

```
Yes, that's way out of date. I can't find "registry" or "archetyhpical" or "get_value_mut" in the current sources, so it appears that this code has been replaced. 

Waiting for the new re-factored release to be announced. I've built test programs with the top of "trunk" but have not modified my main program pending the completion of the current rework.

**Comment by John-Nagle** [comment]
thread 'main' panicked at 'no entry found for key', /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/a972b0f/rend3/src/util/registry/archetypical.rs:126:28
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_display
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:72:5
   3: core::panicking::panic_str
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:56:5
   4: core::option::expect_failed
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:1874:5
   5: rend3::util::registry::archetypical::ArchetypicalRegistry<K,V,HandleType>::get_value_mut
   6: rend3::renderer::ready::ready
   7: rend3::renderer::Renderer::ready
   8: <sharpview::AppUi as rend3_framework::App>::handle_event
   9: rend3_framework::async_start::{{closure}}::{{closure}}
  10: winit::platform_impl::platform::x11::EventLoop<T>::run
  11: winit::platform_impl::platform::EventLoop<T>::run
  12: winit::event_loop::EventLoop<T>::run
  13: rend3_framework::winit_run
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  15: rend3_framework::start
  16: sharpview::main

Happened again. No idea why. Program had been running OK for about 15 minutes.

**Comment by John-Nagle** [comment]
Similar error in the new version (rev = "6dd847f") of Rend3:
```

thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/managers/object.rs:357:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: core::option::Option<T>::unwrap
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:778:21
   4: rend3::managers::object::evaluate::{{closure}}
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/managers/object.rs:357:48
   5: rend3::util::freelist::buffer::FreelistDerivedBuffer::apply::{{closure}}
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/util/freelist/buffer.rs:94:24
   6: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:629:13
   7: core::option::Option<T>::map
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:925:29
   8: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/iter/adapters/map.rs:103:9
   9: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/iter/adapters/enumerate.rs:46:17
  10: rend3::util::scatter_copy::ScatterCopy::execute_copy
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/util/scatter_copy.rs:108:28
  11: rend3::util::freelist::buffer::FreelistDerivedBuffer::apply
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/util/freelist/buffer.rs:101:9
  12: rend3::managers::object::ObjectManager::evaluate
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/managers/object.rs:167:13
  13: rend3::renderer::eval::evaluate_instructions
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/renderer/eval.rs:174:5
  14: rend3::renderer::Renderer::evaluate_instructions
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3/src/renderer/mod.rs:432:9
  15: <sharpview::AppUi as rend3_framework::App>::handle_event
             at /home/john/projects/sl/SL-test-viewer/sharpview/src/main.rs:517:39
  16: rend3_framework::async_start::{{closure}}::{{closure}}
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3-framework/src/lib.rs:332:9
  17: winit::platform_impl::platform::sticky_exit_callback
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/platform_impl/linux/mod.rs:884:9
  18: winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/platform_impl/linux/x11/mod.rs:375:21
  19: winit::platform_impl::platform::x11::EventLoop<T>::run_return
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/platform_impl/linux/x11/mod.rs:483:27
  20: winit::platform_impl::platform::x11::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/platform_impl/linux/x11/mod.rs:498:25
  21: winit::platform_impl::platform::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/platform_impl/linux/mod.rs:792:56
  22: winit::event_loop::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.28.1/src/event_loop.rs:305:9
  23: rend3_framework::winit_run
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3-framework/src/lib.rs:167:5
  24: rend3_framework::async_start::{{closure}}
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3-framework/src/lib.rs:292:5
  25: pollster::block_on
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/pollster-0.3.0/src/lib.rs:128:15
  26: rend3_framework::start
             at /home/john/.cargo/git/checkouts/rend3-e03f89403de3386a/6dd847f/rend3-framework/src/lib.rs:417:9
  27: sharpview::main
             at /home/john/projects/sl/SL-test-viewer/sharpview/src/main.rs:667:5
  28: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

As before, everything is running well, good images on screen for several minutes, then a crash. Failure is, as above, in "archtype" management:

```
 let data_vec = archetype
        .data_vec
        .downcast_slice::<Option<InternalObject<M>>>()
        .unwrap();

    archetype
        .buffer
        .apply(device, encoder, scatter, |idx| data_vec[idx].as_ref().unwrap().inner)
```

The "unwrap()" failed.

**Comment by John-Nagle** [comment]
Same crash four times now. Time to failure is 3-10 minutes. For some reason, there's a "None" in data_vec's array of InternalObject, and the index with the null was referenced, but I have no idea what that means.\

**Comment by John-Nagle** [comment]
This is bad. Something is going wrong deep down in Rend3/WGPU storage management, but I don't know what. Crashes every few minutes.

**Comment by John-Nagle** [comment]
Not sure yet, but I think the problem has to do with object movement of existing objects, done using **set_object_transform** from a thread different than the one that created or is rendering the object. When I comment out object movement in my code, the crashes stop.

Now that I think about it, I never got crashes of this type before I implemented object movement. Once I added it, I started getting crashes very frequently. If I turn movement off, crashes stop. Race condition?

**Comment by John-Nagle** [comment]
It's looking like a race condition. I put a 3ms delay in the render thread after triggering the thread that does **set_object_transform** for moving objects. That gives the movement updates time to complete. I've tried this several times with and without that delay. Without the 3ms delay, it crashes at that panic every time within seconds of startup. With the delay, it usually doesn't crash, although when there are more moving objects, it has crashed after several minutes. So this is very timing-dependent.

Here's my code for the frame update event. I added that one line with a 3ms "sleep", which gives the thread unblocked by "frame_update" time to do its **set_object_transform** calls for moving objects before the render thread gets going again.

```
                //  Start up movement updater for next frame.
                match &mut data.runner {
                    Runner::Replay(_) => {}
                    Runner::Run(run) => run.frame_update(data.frame_time),
                    Runner::None => {} // none, nothing to do
                }
                std::thread::sleep(std::time::Duration::from_millis(3));    // ***TEMP TEST*** let movement finish.
                profiling::scope!("3D");
                // Get a frame
                let frame = surface.unwrap().get_current_texture().unwrap();
                
                // Lock the routines
                let pbr_routine = rend3_framework::lock(&routines.pbr);
                let tonemapping_routine = rend3_framework::lock(&routines.tonemapping);
                let mut skybox_routine = rend3_framework::lock(&routines.skybox);
                // Swap the instruction buffers so that our frame's changes can be processed.
                renderer.swap_instruction_buffers();
                // Evaluate our frame's world-change instructions
                let mut eval_output = renderer.evaluate_instructions();
                
                skybox_routine.evaluate(renderer);
                //  Ready up the routines

                // Build a rendergraph
                let mut graph = rend3::graph::RenderGraph::new();

                // Import the surface texture into the render graph.
                let frame_handle = graph.add_imported_render_target(
                    &frame,
                    0..1,
                    rend3::graph::ViewportRect::from_size(resolution),
                );

                // Add the default rendergraph without a skybox
                base_rendergraph.add_to_graph(
                    &mut graph,
                    &eval_output,
                    &pbr_routine,
                    Some(&skybox_routine),
                    &tonemapping_routine,
                    frame_handle,
                    resolution,
                    SAMPLE_COUNT,
                    glam::Vec3::splat(AMBIENT_LEVEL).extend(1.0), // ambient color
                    glam::Vec4::new(0.10, 0.05, 0.10, 1.0), // Purple. If seen, we have a problem.
                );

                // Add egui on top of all the other passes
                data.egui_routine
                    .add_to_graph(&mut graph, input, frame_handle);

                // Dispatch a render using the built up rendergraph!
                graph.execute(renderer, &mut eval_output);
                // Present the frame
                frame.present();
```

**Comment by John-Nagle** [comment]
Behavior of moving objects can be very weird. See attached image. There's a tram car running along that track. Usually it appears in the scene. Here, the image shows the tram car's shadow, but not the tram car itself. A panic in object.rs at the same "unwrap()" happened a few seconds after this screenshot was taken.

![tramshadowonly](https://user-images.githubusercontent.com/942608/221430828-0b8178b7-3640-4970-bd0c-e010d062be1a.png)

**Comment by John-Nagle** [comment]
Any progress on this? At least a workaround by locking something to prevent the race condition? I'm stalled right now, and can't do anything on my project.

**Comment by cwfitzgerald** [comment]
Right now the main blocker is either information or a repro case. Ideally if we could figure out some way to modify the examples with enough off-thread object spam to reproduce this. Short of that, I was thinking of a way to implement instrumentation such that we might be able to figure out the cause and effect causing the crash.

I have pushed onto a branch called `object-lifespan-logging` which adds some logs in tactical places which should help give an idea of what is going on - it may not find the solution, but it should give us some leads. Run that branch with your code and some way of getting `rend3`'s `INFO` level logs into an text file and the last few seconds of logs before the crash might be able to give us something.

**Comment by John-Nagle** [comment]
Thanks. Let me see what I can do with the branch.

If this has to be reproduced, I'll try taking my "render-bench" city test and adding some moving objects that are moved
from another thread. 

It takes a complex scene to hit this. I can log into simple test areas and run for long periods, but that steampunk city causes a crash quickly.

**Comment by John-Nagle** [comment]
Trying to build branch "object-lifespan-logging". That uses wgpu 0.14 instead of the current 0.15,  an earlier version of egui, etc. Is that intentional? Trying to use that version of Rend (git clone from Github, git checkout object-lifespan-logging") I get errors building my own program as Rend3 compiles:

```
   Compiling rend3 v0.3.0 (/home/john/projects/rend3/rend3)
error[E0560]: struct `Limits` has no field named `max_bindings_per_bind_group`
  --> /home/john/projects/rend3/rend3/src/setup.rs:85:5
   |
85 |     max_bindings_per_bind_group: 640,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Limits` does not have this field
   |
   = note: available fields are: `max_texture_dimension_1d`, `max_texture_dimension_2d`, `max_texture_dimension_3d`, `max_texture_array_layers`, `max_bind_groups` ... and 23 others

error[E0560]: struct `Limits` has no field named `max_bindings_per_bind_group`
   --> /home/john/projects/rend3/rend3/src/setup.rs:117:5
    |
117 |     max_bindings_per_bind_group: 640,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Limits` does not have this field
    |
    = note: available fields are: `max_texture_dimension_1d`, `max_texture_dimension_2d`, `max_texture_dimension_3d`, `max_texture_array_layers`, `max_bind_groups` ... and 23 others

error[E0599]: no method named `unwrap` found for struct `QueueWriteBufferView<'_>` in the current scope
   --> /home/john/projects/rend3/rend3/src/managers/mesh.rs:130:18
    |
130 |                 .unwrap();
    |                  ^^^^^^ help: there is a method with a similar name: `swap`

error[E0599]: no method named `unwrap` found for struct `QueueWriteBufferView<'_>` in the current scope
   --> /home/john/projects/rend3/rend3/src/managers/mesh.rs:140:14
    |
140 |             .unwrap();
    |              ^^^^^^ help: there is a method with a similar name: `swap`

error[E0599]: no method named `unwrap` found for struct `QueueWriteBufferView<'_>` in the current scope
  --> /home/john/projects/rend3/rend3/src/util/buffer.rs:64:72
   |
64 |         let mut mapped = queue.write_buffer_with(&self.inner, 0, size).unwrap();
   |                                                                        ^^^^^^ help: there is a method with a similar name: `swap`

```

**Comment by cwfitzgerald** [comment]
No, that wasn't intentional - was trying to pick the right commit to base that branch off of, and obviously picked wrong - you should be able to cherry pick the commit with the logs onto whatever branch you use, otherwise I can rebase it in the morning

**Comment by John-Nagle** [comment]
OK. I see [the commit](https://github.com/BVE-Reborn/rend3/compare/trunk...object-lifespan-logging). It's better if we're working from identical versions than for me to patch that in somewhere. So I will wait for you to do that tomorrow and tell me which commit to work from. Thanks.

**Comment by cwfitzgerald** [comment]
Branch updated to `93874495f9da228feab71bb429d9bd82bb380d81`

**Comment by John-Nagle** [comment]
I can reproduce the bug and get a log, but the log is mostly full of my own INFO level messages interspersed with yours. Too much is happening in other threads. Gigabytes of log info.
I can barely find the panic info. I'll work on configuring the logging. Meanwhile, here's a new picture from when it works.
![cafenodicawalk](https://user-images.githubusercontent.com/942608/223549780-9a15f85b-f931-4e03-806f-7706eeb0aacc.png)
Enjoy.

**Comment by John-Nagle** [comment]
[rend3bufcrashclean.zip](https://github.com/BVE-Reborn/rend3/files/10913931/rend3bufcrashclean.zip)
Clean log of crash. Panic begins at line 71472.

**Comment by John-Nagle** [comment]
I was looking at the code for **set_transform** in Rend3:

```
#[track_caller]
    pub fn set_object_transform(&self, handle: &ObjectHandle, transform: Mat4) {
        self.instructions.push(
            InstructionKind::SetObjectTransform {
                handle: handle.get_raw(),
                transform,
            },
            *Location::caller(),
       
```
Having a **set_transform** operation on the instruction queue only seems to keep a "raw" copy of the handle. So,  in the following sequence of events:

- User of Rend3 calls **set_object_transform** for object X, which creates an instruction with a raw handle and pushes it on the queue.
- User of Rend3 drops the user's handle to object X, which would seem to delete it if nothing else is going on.
- Frame draw starts and the instruction list is processed, referencing a raw handle to a dead object.

what happens?

Looking at the log I sent previously, note object numbers 2437 and 2438. They both show a **set_transform** followed shortly thereafter by a **drop.**, followed shortly thereafter by a panic.

```
[INFO  rend3::managers::object] Object 2437 update gpu buffer
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/john/projects/rend3/rend3/src/managers/object.rs:363:9
```
It's panicking because object #2437 is invalid. 

I'm not sure I'm reading the sequence of events right, but this looks suspicious.

Maybe **set_transform** instructions need to be an Rc-type owner of the object handle while the instruction is enqueued, so the object can't disappear while the instruction is pending.

Am I reading this correctly?

**Comment by cwfitzgerald** [comment]
Okay I think I know the issue here - and I should have a fix here shortly.

The scenario you listed is valid because the drop impl doesn't directly delete the object just make another instruction to delete it. As set_object_transform has completed pushing the instruction onto the instruction list, and it takes a reference to an owning handle, the delete instruction must be added after the set transform instruction, and we do see this in this scenario.  

The issue is fundamentally that the code shouldn't have the failure be fatal. The order of operations ends up being

- `set_object_transform`
- `remove`
- `ready`

inside the ready call, the code is trying to get the updated transform (as the buffer manager thing tells it that the transform has been updated). it tries to get the data and fails. In a future version of the code (which I showed off at the demo), this is more obviously wrong, as that code uses temporal tricks and defers object deletion.

I pushed a commit `9eaa061001393a83402f1976ad3d22469c5369c6` on branch `fix-465`. Hopefully this should fix it for you.

**Comment by John-Nagle** [comment]
So far, so good. No crashes. Big improvement.

That strange bug where, for moving objects, the object disappears but it still continues to cast a shadow remains. No idea what's causing that. It's not a new bug; I just see it now because I have more motion support. More on that as a new issue when I have more info. I think it may be related to changing texture resolutions.

But I'm back in business now. I can run my program.
